### PR TITLE
Store file path and file id in token stream metadata

### PR DESCRIPTION
**Pull request by maciektr**
commit-id:98fd5a13

---

**Stack**:
- #1200
- #1195
- #1194
- #1192 ⬅

⚠️ *Part of a stack created by [spr](https://github.com/ejoffe/spr). Do not merge manually using the UI - doing so may have unexpected results.*

### CODE REVIEW

File: plugins/cairo-lang-macro-stable/src/lib.rs [hunk 1]
```diff
@@ -1,24 +1,44 @@
...
+/// An option.
+///
+/// This struct implements FFI-safe stable ABI.
+#[repr(C)]
+#[derive(Debug)]
+pub enum StableOption<T> {
+    None,
+    Some(T),
+}
+
```
Comment (mkaput) [inline plugins/cairo-lang-macro-stable/src/lib.rs:17]: You're using this to store pointers. Can't you put `NonNull` there and use standard Option instead?

https://doc.rust-lang.org/stable/std/option/index.html#representation

**Review by mkaput — CHANGES_REQUESTED** [review]

File: scarb/tests/build_cairo_plugin.rs [hunk 2]
```diff
@@ -530,7 +530,50 @@ fn can_return_aux_data_from_plugin() {
...
+    let temp = TempDir::new().unwrap();
+    let t = temp.child("some");
+    simple_project_with_code(
+        &t,
+        indoc! {r##"
+        use cairo_lang_macro::{ProcMacroResult, TokenStream, attribute_macro};
+
+        #[attribute_macro]
+        pub fn some_macro(token_stream: TokenStream) -> ProcMacroResult {
+            println!("{:?}", token_stream.metadata());
```
Comment (mkaput) [inline scarb/tests/build_cairo_plugin.rs:547]:
```suggestion
            println!("{:#?}", token_stream.metadata());
```

and you'll get much easier to diff output

File: plugins/cairo-lang-macro/src/lib.rs [hunk 3]
```diff
@@ -66,18 +69,59 @@ pub enum ProcMacroResult {
...
 #[derive(Debug, Default, Clone)]
-pub struct TokenStream(String);
+pub struct TokenStream {
+    value: String,
+    metadata: TokenStreamMetadata,
+}
+
+#[derive(Debug, Default, Clone)]
+pub struct TokenStreamMetadata {
+    original_file_path: Option<String>,
```
Comment (mkaput) [inline plugins/cairo-lang-macro/src/lib.rs:79]: Do we need file path if we have fileid ?

We wanted to include paths in spans anyway, and in this context both fields provide the same value

File: plugins/cairo-lang-macro/src/lib.rs [hunk 4]
```diff
@@ -66,18 +69,59 @@ pub enum ProcMacroResult {
...
+            s.with_file_path(file_path)
+        } else {
+            s
+        }
+    }
+
+    #[doc(hidden)]
+    pub fn with_file_path(mut self, file_path: impl ToString) -> Self {
+        self.original_file_path = Some(file_path.to_string());
+        self.file_id = Some(short_hash(file_path.to_string()));
```
Comment (mkaput) [inline plugins/cairo-lang-macro/src/lib.rs:123]: shouldn't this hash be computed inside scarb?

**Review by mkaput — CHANGES_REQUESTED** [review]

File: plugins/cairo-lang-macro/src/lib.rs (hunk 3 above)
Comment (maciektr) [reply plugins/cairo-lang-macro/src/lib.rs:79]: Discussed offline

**Review by mkaput — APPROVED** [review]

**Pull request merged**